### DISCUSSION TITLE
Should the 'exit-function' method be triggered when using 'completion-at-point'?

### DISCUSSION THREAD

**Discussion by jadestrong** [minad/corfu#491] [Q&A] [answered]
![QuickTime movie](https://github.com/user-attachments/assets/8e03b10d-1baa-4772-ac63-2a4ad665c28f)

Here are two items with the same name, but with different parameters. When the 'completion-at-point' method is triggered using the Tab key, it completes the common prefix string, but does not trigger the custom 'exit-function' defined by capf).

https://github.com/minad/corfu/blob/5e3a959766d2313651c5db3beedd937bfc27b57a/corfu.el#L925

**Comment by minad** [comment]
If there is only a single item after common prefix completion, the exit function will be called, see: 
https://github.com/minad/corfu/blob/5e3a959766d2313651c5db3beedd937bfc27b57a/corfu.el#L940

Otherwise Corfu should show the popup after completing the prefix, such that the exit function should be called later on, after completion finished.

The problem here is that Corfu doesn't detect the single candidate, since you have two candidates with the same name, right? We could use a different check to handle duplicates. Instead of the `if` conditional:

https://github.com/minad/corfu/blob/5e3a959766d2313651c5db3beedd937bfc27b57a/corfu.el#L931

Use this `if-let`:

~~~ emacs-lisp
(if-let ((single (car candidates))
         ((cl-loop for c in candidates always (equal c single))))
    ...
  ...)
~~~

Then the problem is that the exit function won't know which one of the two candidates has actually been selected, since the selection wasn't explicit. But the exit function of the Capf will expand the first candidate then, which would be the reasonable thing to do. 

Can you please check if the alternative `if-let` gives the desired result in this case?

**Comment by minad** [comment]
> Here are two items with the same name, but with different parameters. When the 'completion-at-point' method is triggered using the Tab key, it completes the common prefix string, but does not trigger the custom 'exit-function' defined by capf).

Thinking about this more, I suspect I've misunderstood the problem. After common prefix completion, the exit function should not be called, as long as the candidate is not fully completed. The exit function should only be called when completion finished resulting in a full candidate. So in this particular case, the answer to your question is "no".

**Reply by jadestrong** [reply]
When `completion-try-completion` receives two completion items with the same label, it automatically completes the common prefix `effects?`.  But here we actually need to use `newText` in `textEdit`. Directly using `label` to complete is not as expected, as it has an extra `?`, it's not necessary. Is there a solution to this?

```
[
    CompletionItem {
        item: CompletionItem {
            label: "effects?",
            label_details: Some(
                CompletionItemLabelDetails {
                    detail: None,
                    description: None,
                },
            ),
            kind: Some(
                Field,
            ),
            detail: None,
            documentation: None,
            deprecated: None,
            preselect: None,
            sort_text: Some(
                "12\0effects\0",
            ),
            filter_text: Some(
                "effects",
            ),
            insert_text: Some(
                "effects",
            ),
            insert_text_format: Some(
                PlainText,
            ),
            insert_text_mode: None,
            text_edit: Some(
                Edit(
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 10,
                                character: 2,
                            },
                            end: Position {
                                line: 10,
                                character: 4,
                            },
                        },
                        new_text: "effects",
                    },
                ),
            ),
            additional_text_edits: None,
            command: Some(
                Command {
                    title: "",
                    command: "_vtsls.completionCacheCommand",
                    arguments: Some(
                        [
                            Object {
                                "cacheId": Number(6),
                                "index": Number(0),
                                "providerId": Number(2),
                            },
                        ],
                    ),
                },
            ),
            commit_characters: Some(
                [
                    ".",
                    ",",
                    ";",
                ],
            ),
            data: Some(
                Object {
                    "cacheId": Number(6),
                    "index": Number(0),
                    "providerId": Number(2),
                },
            ),
            tags: None,
        },
        language_server_id: Some(
            0,
        ),
        language_server_name: "vtsls",
        start: 206,
        end: 212,
    },
    CompletionItem {
        item: CompletionItem {
            label: "effects?",
            label_details: Some(
                CompletionItemLabelDetails {
                    detail: Some(
                        "(form)",
                    ),
                    description: None,
                },
            ),
            kind: Some(
                Field,
            ),
            detail: None,
            documentation: None,
            deprecated: None,
            preselect: None,
            sort_text: Some(
                "12\0effects\01",
            ),
            filter_text: Some(
                "effects(form) {\n    $0\n},",
            ),
            insert_text: Some(
                "effects(form) {\n    $0\n},",
            ),
            insert_text_format: Some(
                Snippet,
            ),
            insert_text_mode: None,
            text_edit: Some(
                Edit(
                    TextEdit {
                        range: Range {
                            start: Position {
                                line: 10,
                                character: 2,
                            },
                            end: Position {
                                line: 10,
                                character: 4,
                            },
                        },
                        new_text: "effects(form) {\n    $0\n},",
                    },
                ),
            ),
            additional_text_edits: None,
            command: Some(
                Command {
                    title: "",
                    command: "_vtsls.completionCacheCommand",
                    arguments: Some(
                        [
                            Object {
                                "cacheId": Number(6),
                                "index": Number(1),
                                "providerId": Number(2),
                            },
                        ],
                    ),
                },
            ),
            commit_characters: Some(
                [
                    ".",
                    ",",
                    ";",
                ],
            ),
            data: Some(
                Object {
                    "cacheId": Number(6),
                    "index": Number(1),
                    "providerId": Number(2),
                },
            ),
            tags: None,
        },
        language_server_id: Some(
            0,
        ),
        language_server_name: "vtsls",
        start: 206,
        end: 212,
    },
]
```

**Reply by minad** [reply]
I am not sure what you are asking exactly, but I've tested Corfu again using the following simple `test-capf`. It seems to work correctly for the duplicate "word" (note the different text properties). When I enter `wo TAB` it completes the prefix to "word" and displays the popup, such that I can select among the duplicate candidates. The other *exact* duplicate candidate "wand" is filtered out and treated as a single candidate only.

~~~ emacs-lisp
(defun test-capf ()
  (when-let ((bounds (bounds-of-thing-at-point 'symbol)))
    (list (car bounds) (cdr bounds)
          (list (propertize "word" 'face 'italic)
                "word" "wand" "wand" "wind")
          :exit-function (lambda (&rest args) (message "EXIT %S" args)))))

(setq-local completion-at-point-functions '(test-capf)
            completion-cycle-threshold nil)
~~~

**Reply by jadestrong** [reply]
Thank you very much for your assistance. I have tested with the default configuration, and it works as you described. It's possible that the doomemacs configuration or my personal configuration is causing this issue. I will investigate further to identify the problem. Thank you once again for your support.

**Reply by jadestrong** [reply]
I reproduced the issue through the following step:

1. use this `test-capf`
 ```
(defun test-capf ()
  (when-let ((bounds (bounds-of-thing-at-point 'symbol)))
    (list (car bounds) (cdr bounds)
          (list (propertize "word?" 'face 'italic)
                "word?" "wand" "wand" "wind"))))
```
2. enable `js-mode` or `typescript-mode` or `typescript-ts-mode`
3. set local variable 
```
(setq-local completion-at-point-functions '(test-capf)
            completion-cycle-threshold nil)
```
4. enter `wo`  + `Tab` will only completes prefix to `word?` and no popup display.

**Reply by minad** [reply]
I see. The problem seems to be that `word?` is not recognized as a symbol or word at point. This means that right after TAB, completion terminates immediately since the token before point is not recognized as it should. The behavior is similar to when you are entering another terminating character, e.g., a semicolon or a brace.

You can try this by calling `(thing-at-point 'symbol)` or `(thing-at-point 'word)` right after `word?`. However if you put the point directly behind `word`, then `thing-at-point` will work as expected. 

As far as I know, the question mark is not a valid symbol character in JavaScript, so it is a little odd that it is produced by the Capf or Lsp. However from your code it looks like question mark may be a valid symbol character in TS? Then this should be fixed in the syntax table of the TS major mode. This is quite a special edge case you've found here.

**Reply by jadestrong** [reply]
The question mark is not a valid symbol character in TypeScript either.  `effects?`'s `?` indicates that the`effects` field is optional, not required. Since the Language Server Protocol (LSP) uses the `insertText` or `textEdit` field as the correct content for insertion, it needs the `exit-function` to execute the correct insertion logic (remove the inserted label and replace it with the `insertText` instead). It's just that in this particular scenario,  `completion-try-completion`  happens to complete the full label.

**Reply by jadestrong** [reply]
I tried company-mode, and it also has this problem, maybe there is no good solution. We'll have to leave it like this for now. Thanks.

**Reply by minad** [reply]
I understand. I consider this a bug in the LSP Capf, since it generates an invalid symbol as an intermediate step, but at the same time doesn't recognize it as a symbol, such that completion unexpectedly stops. There are two options for a fix on the side of the Capf - either don't generate symbols with question marks or recognize the boundaries of symbols with question marks.

**Reply by jadestrong** [reply]
Actually, this is because the LSP's completion items are not compatible with Emacs's string-based completion mechanism. LSP only uses the label to display more information to make selection easier, and it uses the `insertText` or `textEdit` field for insertion, the `filterText` field for filtering, and so on. To adapt to Emacs's style, `lsp-mode` and `eglot` use the `label` as a string and set the completion item as a text property, which causes this problem.
I would like to consult on whether it's possible to support  an alternative option where, when enter `Tab`, it doesn't attempt to complete the common prefix but instead directly show a popup to facilitate selection.

**Reply by minad** [reply, marked as answer]
> Actually, this is because the LSP's completion items are not compatible with Emacs's string-based completion mechanism.

The right thing would probably be to use `insertText` as candidate string, if available and fallback to `label` if unavailable. But this has other downsides of course.

> I would like to consult on whether it's possible to support an alternative option where, when enter Tab, it doesn't attempt to complete the common prefix but instead directly show a popup to facilitate selection.

Such an option will deviate too much from usual completion behavior, e.g., in shell. However I believe you can implement such non-expansion behavior with a custom completion style.

**Reply by jadestrong** [reply]
> > Actually, this is because the LSP's completion items are not compatible with Emacs's string-based completion mechanism.
> 
> The right thing would probably be to use `insertText` as candidate string, if available and fallback to `label` if unavailable. But this has other downsides of course.
> 
Yes. `insertText` cannot be guaranteed to be a normal string, such as `effects(form) {\n    $0\n}` , it will expand as a snippet.
> > I would like to consult on whether it's possible to support an alternative option where, when enter Tab, it doesn't attempt to complete the common prefix but instead directly show a popup to facilitate selection.
> 
> Such an option will deviate too much from usual completion behavior, e.g., in shell. However I believe you can implement such non-expansion behavior with a custom completion style.

Got it , I will investigate it. Thanks for your reply.

**Reply by minad** [reply]
> Yes. insertText cannot be guaranteed to be a normal string, such as effects(form) {\n    $0\n} , it will expand as a snippet.

Honestly, this whole LSP design is a mess. Emacs in contrast seems straightforward, despite all its baroque legacy.

Another option would be to handle the problem on the symptom level. We could try to make sure that the exit function is guaranteed to be called, even when bailing out due to the completion predicate. This means adding an additional condition right after `corfu--replace`/`goto-char`. Check the predicate and if it fails, immediately call the exit function with status finished without setting up Corfu. (EDIT: But then we lose the ability to select among duplicates. So it would not be the correct thing to do. It would be equally wrong as my initial proposal in https://github.com/minad/corfu/discussions/491#discussioncomment-10333840. The problem is that showing the popup requires the completion predicate to succeed. However you can set `corfu-quit-at-boundary=nil` but this will likely lead to other unexpected behavior.)

Could you please check how default completion behaves in this case after pressing TAB? Does it call always call the exit function, even when the completion predicate fails due to the question mark?

**Reply by jadestrong** [reply]
Yes，I think the simplest solution is not to try to complete the common prefix. This can avoid unexpected characters causing the predicate to fail, and it can allow selection among duplicates. To accommodate similar cases in the current implementation is difficult, and it may cause other unknown problems.
